### ISSUE TITLE
Extremely high CPU usage for headless chrome

### ISSUE THREAD

**Issue by saifuddin** [open-wa/wa-automate-nodejs#1345]
So, I'm not seeing anyone else having a problem with this, but I'm at my wit's ends, so posting this here is my last resort I guess 😁 . I'm seeing extremely high and constant CPU usage overtime which I tracked to a headless chrome.

**Describe the bug**
Once QR has been scan (WhatsApp session started), CPU usage by headless chrome jumps quite high, 20-80% of CPU load.

**Steps to Reproduce**
1. Execute below sample code
2. Scan QR code

**create() code**
```javascript
var wa = require('@open-wa/wa-automate');

var _waConfig = {
	sessionId: '<your_phone>',
	licenseKey: '<your_license>',
	authTimeout: 0
};

var _whatsappStart = (whatsapp) => {
	console.log('Successfully start WhatsApp Web');

	// ==============
	// WHATSAPP EVENT
	// ==============

	// returns 'CONNECTED' or 'TIMEOUT' or 'CONFLICT' (if user opens whatsapp web somewhere else)
	whatsapp.onStateChanged((state) => {
		console.log('onStateChanged', state);

		switch (state) {
			case 'CONFLICT':
			case 'UNLAUNCHED':
				whatsapp.forceRefocus();
				break;
			default:
				break;
		}
	});

	whatsapp.onAnyMessage((message) => {
		console.log(`=== onAnyMessage - ${message.type} ===`);
		console.log(JSON.stringify(message));
		console.log('=== END ===');
	});
};

wa.create(_waConfig)
.then(async client => await _whatsappStart(client))
.catch(e => {
	console.log('CREATE ERROR', e.message);
});
```

**Expected behavior**
CPU usage should not constantly be on high

**DEBUG INFO**
```
Debug Info {
  WA_VERSION: 2.2104.9,
  PAGE_UA: WhatsApp/2.2037.6 Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36,
  WA_AUTOMATE_VERSION: 3.4.0,
  BROWSER_VERSION: HeadlessChrome/90.0.4403.0,
  OS: Linux 5.4
}
```

**Screenshots**
![image](https://user-images.githubusercontent.com/175973/107842229-2abd6580-6d87-11eb-8f6d-206dd1716316.png)

**Host**
 - OS: Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-1037-aws x86_64)

**Additional context**
None

Anybody else having this issue? Or possibly only on Ubuntu Server 20.04? Or is this something I have to bring over to the puppeteer (https://github.com/puppeteer/puppeteer)?

Thanks!

**Comment by smashah** [comment]
@saifuddin exquisite use of the template. Much appreciated!

Let's wait for some corroboration of this issue. My hunch is that it is due to puppeteer itself since there hasn't really been any changes in regards to the libraries use of puppeteer for a while. I may be missing some optimization though so I'll check on this.

In the meantime can you try with `useChrome: true` and see if it makes a difference to your memory use. 

Also, what environment are you running this on? Local machine or Heroku/was/etc? Whats the source of this memory usage chart?


Thanks

**Comment by saifuddin** [comment]
@smashah Thanks for your reply, and apologies for my late reply! I was trying out a few stuff on my end, but it seems like everything I did, the CPU usage was still very high 😅 

> Let's wait for some corroboration of this issue

Sure thing! It's still very weird tho, that I haven't seen anyone complaining - so, I'm not really sure what I did wrong.

> My hunch is that it is due to puppeteer itself

Yeah, I suspect that's the case; altho I'm not sure where to debug puppeteer itself. I did came across a SO question regarding a similar issue (https://stackoverflow.com/questions/49008008/chrome-headless-puppeteer-too-much-cpu/) and also to change the setting to this (https://github.com/open-wa/wa-automate-nodejs/blob/707ee87d8311b2a9de88455ffe64266fdf7a31c7/src/config/puppeteer.config.ts - with the commented flags) - but, still no luck.

My CPU usage starts at ~18% after I scanned the QR Code and continued to climb until it steadied at around 70-80%.

> In the meantime can you try with useChrome: true and see if it makes a difference to your memory use.

Tried that, but the CPU usage still is the same.

Funnily enough, the memory usage is not an issue; when my CPU usage jumped and stayed high; memory usage is the same as before the CPU jump

![image](https://user-images.githubusercontent.com/175973/107952191-0c11c700-6f5f-11eb-8948-0f37bb6a5fdb.png)
![image](https://user-images.githubusercontent.com/175973/107952240-1c29a680-6f5f-11eb-8815-965b1ff4fb78.png)

> Also, what environment are you running this on? Local machine or Heroku/was/etc?

I've replicated this on a few platforms actually. On AWS Free tier (t2.micro), on DigitalOcean Basic Droplet (1GB RAM) and so far all have the same issue - high CPU usage. It was running fine on these small VPSes before (from the graph)

> Whats the source of this memory usage chart?

Sorry if I misunderstood the question - are you referring to how I generate the chart? I use HetrixTools

I have my VPS open for anyone who wants to see it tho 😝 

Thanks again @smashah !

**Comment by allanmelo** [comment]
TL;DR: try to use a more powerful instance. An instance in Google Cloud with 2 cores and 2 GB of RAM proved to be ok in an application that I kept for a long time.

-----

Just to corroborate what @saifuddin said, I also detected a high CPU usage when I tried to create a dedicated instance of wa-automate-nodejs for a bot I was managing.

I first detected the problem while running it on an AWS t2-micro instance. The CPU was operating at 110%, which made me penalized for the excessive use of Amazon's "burst time", making the whole application extremely slow. I only detected the problem using AWS's own monitoring tools. Usually, I monitor this using the PM2, but the PM2 monitor only showed processing spikes when there was some kind of request coming from use on WP.

Before that, I ran this same application on a Raspberry Pi 2b, with Ubuntu Server 20.04. The Pi2b seems to run at the limit, overheating over time (but this was solved with a little fan over it). It ran stable after that for a long time, staying weeks without even needing a reboot (but I just ended up doing some reboots because of some updates in the application)

I recently tried to put it on a more powerful instance on Google Cloud. I configured a machine with 2 cores and 2 GB of Ram. It worked very well for a while, but I haven't more details to provide about that since I'm having some trouble running wa-automate-nodejs in any environment for a while. Don't know what's happen & unfortunately I'm with so little time to go deep on this.

**Comment by smashah** [comment]
@allanmelo @saifuddin it's been a crazy week and a bit. Recently there was a major bump in the pptr deps. Have you tried the latest version of the library? Are you still seeing these CPU issues?

**Comment by smashah** [comment]
@allanmelo @saifuddin Today I had a video call with another user who was experiencing this problem where the CPU was being pinned to 100% shortly after starting the process. The problem was resolved by updating the library to the latest version (now 3.9.2).

I highly encourage everyone to keep up to date with the library at all times.

I will close this issue for now but feel free to bump if you continue to experience this problem.

Thanks for the in-depth and extensive discussion/comments - much more useful than most other issues.
